### PR TITLE
Allow extensions to have external dependencies

### LINKED ISSUE #839: Extension should be able to have external depedencies
**What would you like to be added**:

Logic that handles installation of extension dependencies.

**Why is this needed**:

Extension might need external deps ("there is a npm for everything").

### PR DESCRIPTION
**Pull request by jakolehm**
Fixes #839

### CODE REVIEW

File: src/extensions/extension-manager.ts [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+    fs.ensureDir(path.join(this.extensionPackagesRoot, "node_modules"))
+    fs.writeFileSync(path.join(this.extensionPackagesRoot, "package.json"), `{"dependencies": []}`, {mode: 0o600})
+
+    return await this.loadExtensions();
+  }
+
+  async getExtensionByManifest(manifestPath: string): Promise<InstalledExtension> {
+    let manifestJson: ExtensionManifest;
+    let mainJs: string;
```
Comment (nevalla) [inline src/extensions/extension-manager.ts]: This seems to be redundant now(?)

File: src/extensions/extension-loader.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+// lazy load so that we get correct userData
+export function extensionPackagesRoot() {
+  return path.join((app || remote.app).getPath("userData"))
+}
+
+export function withExtensionPackagesRoot(callback: Function) {
+  const cwd = process.cwd()
+  try {
+    process.chdir(path.resolve(extensionPackagesRoot()))
```
Comment (ixrock) [inline src/extensions/extension-loader.ts]: Why it needs to be resolved (it's already wrapped to `path.join`)?

File: src/extensions/extension-loader.ts [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+          if (!this.getById(ext.manifestPath)) {
+            this.extensions.set(ext.manifestPath, ext)
+          }
+        })
+      })
+    }
+  }
+
+  autoEnableOnLoad(getLensRuntimeEnv: () => LensRuntimeRendererEnv, { delay = 0 } = {}) {
+    logger.info('[EXTENSIONS-MANAGER]: auto-activation loaded extensions: ON');
```
Comment (ixrock) [inline src/extensions/extension-loader.ts]: Maybe `EXTENSION-LOADER` :)?

File: src/extensions/extension-manager.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+  }
+
+  get folderPath(): string {
+    return path.resolve(__static, "../extensions");
+  }
+
+  async load() {
+    logger.info("[EXTENSION-MANAGER] loading extensions from " + this.extensionPackagesRoot)
+
+    fs.ensureDir(path.join(this.extensionPackagesRoot, "node_modules"))
```
Comment (ixrock) [inline src/extensions/extension-manager.ts]: Isn't it should be `await`-ed before writing `package.json`?

File: src/extensions/extension-manager.ts [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        global: false,
+        prefix: this.extensionPackagesRoot,
+        dev: false,
+        spin: false,
+        "ignore-scripts": true,
+        loglevel: "silent"
+      }, (err) => {
+        console.log = function() {
+          // just to ignore ts empty function error
+        }
```
Comment (ixrock) [inline src/extensions/extension-manager.ts:61]: Commented out `origLogger` patching and can't spot the console issue.. 
What it hides/produces anyway? Where it's visible?

File: src/extensions/extension-store.ts [hunk 6]
```diff
@@ -1,13 +1,8 @@
 import type { ExtensionId, ExtensionManifest, ExtensionVersion, LensExtension } from "./lens-extension";
```
Comment (ixrock) [inline src/extensions/extension-store.ts]: Do we still need this file?

File: src/main/index.ts [hunk 7]
```diff
@@ -75,6 +77,8 @@ async function main() {
 
   // create window manager and open app
   windowManager = new WindowManager(proxyPort);
+
+  extensionLoader.extensions.replace(await extensionManager.load())
```
Comment (ixrock) [inline src/main/index.ts:81]: So, currently all extensions only `renderer` specific? Or how to customize, for example, app global menu from extension? (probably out of scope / needs to create separate issue).

**Review by ixrock — APPROVED** [review]
LGTM

File: src/extensions/extension-loader.ts (hunk 2 above)
Comment (jakolehm) [reply src/extensions/extension-loader.ts]: Good point, it's not needed anymore.

File: src/extensions/extension-loader.ts (hunk 3 above)
Comment (jakolehm) [reply src/extensions/extension-loader.ts]: Yes 🤦

File: src/extensions/extension-manager.ts (hunk 4 above)
Comment (jakolehm) [reply src/extensions/extension-manager.ts]: Yes 🤦

File: src/extensions/extension-manager.ts (hunk 1 above)
Comment (jakolehm) [reply src/extensions/extension-manager.ts]: Yes.

File: src/extensions/extension-manager.ts (hunk 5 above)
Comment (jakolehm) [reply src/extensions/extension-manager.ts:61]: This hides `npm install` output (it outputs few lines even if loglevel is `silent`).

File: src/extensions/extension-store.ts (hunk 6 above)
Comment (jakolehm) [reply src/extensions/extension-store.ts]: I left store related bits to this file but maybe we should remove it for now (until we really need it).

File: src/main/index.ts (hunk 7 above)
Comment (jakolehm) [reply src/main/index.ts:81]: I think we can support `main` extensions just by adding a call to `extensionLoader.autoEnableOnLoad()` before this line. Probably should be done in a separate PR.

Comment (nevalla) [reply src/main/index.ts:81]: I'll create issue about this. Played a bit around this today, and it's not that simple.

**Comment by jakolehm** [comment]
@ixrock @nevalla PTAL again

**Review by nevalla — APPROVED** [review]
LGTM

**Pull request merged**